### DISCUSSION TITLE
High-level interface

### DISCUSSION THREAD

**Discussion by mtanneau** [mtanneau/MathOptPresolve.jl#4] [Ideas]
(I'm trying out this new GH feature) 
cc @joehuchette @frapac

**Q: What should the high-level interface look like?**

Some directions:
* Having an interface to MOI
* How will the user interact with pre/post crush?
* What low-level components do we want to expose?

**Comment by joehuchette** [comment]
I would prefer an MOI interface. I would propose that we: keep the existing interface, suitably altered to support IPs and/or to cater to the needs of folks who intend to use it directly. Then, we can build an MOI wrapper directly on top of this.

Why would a user want to interact with the crushes?

Also, having an interface to configure the presolve routines is a priority for me.

Also, cc @dpo

**Reply by mtanneau** [reply]
Not everybody works with MOI, including me within Tulip.
I think @dpo mentioned he would be interested in a lower-level interface.

> Why would a user want to interact with the crushes?

"interact with crushes" may be as simple as having a `precrush` and `postcrush` function.
The question is more about their signature, whether we separate primal & dual crush, etc.
In any case, how the presolved problem is solved (if solved at all) is up to the user. IMO, so should be interacting with crushes.

**Comment by mtanneau** [comment]
For starters:

### MOI interface

I'm thinking we could have a thin wrapper, which would do the following
1. Extract problem data from original MOI model
2. Preform presolve
3. Instantiate a `presolved_model::MOI.ModelLike` that holds the presolved model
4. Return `presolved_model` along with pre- and post-crush functions

The user side may look like
```julia
my_model = generate_some_model()

# Perform presolve; a new MOI.ModelLike is returned
presolved_model, precrush, postcrush = MOP.presolve(my_model)

# Copy presolve_model into an actual optimizer
my_optimizer = XXX.Optimizer()
MOI.optimize!(my_optimizer, presolved_model)

# Perform post-crush: open question
```

Some remarks:
* Calling `MOI.optimize!` on `presolved_model` should error, because in general it does not have an optimizer attached.
  We could relax this, e.g., when an `OptimizerWithAttributes` is provided:
  ```julia
  presolved_model, precrush, postcrush = MOP.presolve(my_model, MOI.OptimizerWithAttributes(Gurobi.Optimizer))
  MOI.optimize!(presolved_model)  # solve with Gurobi directly
  ```

* Pre- and post-crush is performed by the user. I should be possible to have `precrush` and `postcrush` work with MOI indices directly, to make things more convenient. Note that precrush and postcrush should operate on the full solution vector.
* If `my_model` is modified, then the presolved model is no longer valid.
